### PR TITLE
[WIP] Staging 0.7.0

### PR DESCRIPTION
**Pull request by dritter**
Here is the big merge of #771 and #859 . Please have a look if everything is all right.
I did a quick test (just open the shell, and see that everything is all right) and got an exception in the `status` segment. I fixed that with 8a6597def7ba2bda54fe6d695ded7c2c477a9b0e . That is the only additional change I've made on this branch (besides a bit of cleanup in d11aba28c0f3233ae388d25cb2e2534deddc4d3f and fixing the path to the `test_stats` segment).

Biggest pain points in this merge was that code that was moved changed in `next` and git not being able to track that, so I had to do that manually. I went through all segments and all icons to track the code by hand.

I hope I got everything right and did not forgot anything.

Btw. besides a vast change in the code base, my biggest concerns about this are:
- `vi_mode` segment. The changes in #771 originate from #344 and were built for an async environment. I am not sure how this will behave once we have different generators. Currently, @onaforeignshore disabled it for the default generator.
- ~~`vcs` segment. @onaforeignshore enabled queues in mercurial. This needs to be tested as well.~~ Mercurial Queues were removed. See [chat](https://gitter.im/bhilburn/powerlevel9k?at=5b1bf27a659aad4a3262af02). Commit: 3064838b47c4aeda468562b079a3102b4f22aac4
- Loading segments. In `powerlevel9k.zsh-theme` we go through all possible segments and load only the segments in use. This has two problems: `segmentsInUse` does not take joined segments into account. And secondly we could loop over the configured segments and try to load them (This is probably the shorter list, so this is an performance issue).

~~I'll have a look at the tests tomorrow.~~ Tests now pass.

Quick links:
[Compare 0.6.5 and 0.7.0](https://github.com/dritter/powerlevel9k/compare/staging_065...dritter:staging_700)
[Compare #771 and 0.7.0](https://github.com/dritter/powerlevel9k/compare/onaforeignshore:code_separation...dritter:staging_700)

### CODE REVIEW

**Comment by onaforeignshore** [comment]
I'm reviewing this PR over the next few days. Initial missing item I have noticed is custom segments. We still have to decide where this will be. The latest trend is to place them under `~/.config`.

**Comment by onaforeignshore** [comment]
This branch still uses only `registerIcon`, I expanded that into `registerSegment` in #833, which IMHO  is a better structure as it registers everything to do with a segment in one call.

Since #833 is only a test bed, it shouldn't be pulled straight into this if we want to use it, but I am willing to do the recoding to incorporate that into here. Should only take a few days to do.

Do we want to just go all out, or do we want to do it in stages, bearing in mind that this PR will differ from #833 's `registerSegment`, meaning that people will have to update their custom segments again.

Do we want to go for the speed that predetermining all the segment formatting gives us or stick to this in-between version?

@bhilburn @dritter What are your thoughts on this?

**Comment by dritter** [comment]
Haven't had time to look into #833 [Diff-Quicklink](https://github.com/dritter/powerlevel9k/compare/staging_700...onaforeignshore:rewrite).

About the question how to proceed, and what should be in the next Release:
In my opinion we shouldn't release a intermediate version. I'd like to keep the count of breaking changes as little as possible. Therefore, if we already know that some changes will break backwards compatibility, we should merge them in. The changes of this branch are a massive change anyway, so it is a good idea to include the performance optimizations (like `registerSegment`).

But, to avoid having such massive changes all the way, it would be good to create small PRs with one feature at a time. Otherwise it will be reviewing / testing Hell.

Tl;dr:
I would suggest to chop #833 in small separate PRs with this one as base. And to incorporate the performance improvements before releasing a new version.

File: README.md [hunk 1]
```diff
@@ -210,7 +210,7 @@ without affecting the icon color.
 
 You can also change the battery icon automatically depending on the battery
 level. This will override the default battery icon. In order to do this, you
-need to define the `POWERLEVEL9k_BATTERY_STAGES` variable.
+need to define the `P9k_BATTERY_STAGES` variable.
```
Comment (onaforeignshore) [inline README.md]: Shouldn't the "k" be "K"?

Comment (dritter) [reply README.md]: Good catch!

**Comment by dritter** [comment]
Closing in favor of #865 and following PRs.
